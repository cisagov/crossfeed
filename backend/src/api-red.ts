// Main entrypoint for red.
// Essentially, this entrypoint mimics API Gateway by passing down all requests
// to the crossfeed prod api lambda function.
import * as express from 'express';
import * as bodyParser from 'body-parser';
import * as cors from 'cors';
import { Lambda } from 'aws-sdk';

// Name of the lambda function name.
const API_LAMBDA_FUNCTION_NAME = 'crossfeed-prod-api';
const AWS_REGION = 'us-east-1';

interface LambdaResponse {
  isBase64Encoded: boolean;
  statusCode: number;
  headers: { [x: string]: string };
  multiValueHeaders: { [x: string]: string[] };
  body: string;
}

const app = express();
app.use(cors());
app.use(bodyParser.json());

app.use(async (req, res) => {
  // See https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format
  // for input / output formats for API gateway lambda integration.
  let body = JSON.stringify(req.body);
  if (body === '{}') {
    body = '';
  }
  const input = {
    resource: '/{any+}',
    path: req.path,
    httpMethod: req.method,
    headers: req.headers,
    multiValueHeaders: {},
    queryStringParameters: req.query,
    multiValueQueryStringParameters: {},
    pathParameters: {
      any: req.path.substr(1) // Remove leading slash from path
    },
    stageVariables: null,
    requestContext: {},
    body: body,
    isBase64Encoded: false
  };
  const lambdaClient = new Lambda({
    region: AWS_REGION
  });
  const response = await lambdaClient
    .invoke({
      FunctionName: API_LAMBDA_FUNCTION_NAME,
      InvocationType: 'RequestResponse',
      Payload: JSON.stringify(input)
    })
    .promise();
  const lambdaResponse: LambdaResponse = JSON.parse(
    response.Payload!.toString()
  );
  console.log(input.httpMethod, input.path, lambdaResponse.statusCode);
  if (lambdaResponse.statusCode >= 400) {
    console.error(
      'Request failed with status code',
      lambdaResponse.statusCode,
      'input',
      input,
      'response',
      lambdaResponse
    );
  }
  res.status(lambdaResponse.statusCode);
  for (const k in lambdaResponse.headers) {
    res.header(k, lambdaResponse.headers[k]);
  }
  res.send(lambdaResponse.body);
});

const port = 3000;
app.listen(port, () => {
  console.log('App listening on port ' + port);
});
