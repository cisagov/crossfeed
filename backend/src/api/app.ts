import * as express from 'express';
import * as bodyParser from 'body-parser';
import * as cookieParser from 'cookie-parser';
import * as cookie from 'cookie';
import * as cors from 'cors';
import * as helmet from 'helmet';
import { handler as healthcheck } from './healthcheck';
import { handler as scheduler } from '../tasks/scheduler';
import * as auth from './auth';
import * as domains from './domains';
import * as search from './search';
import * as vulnerabilities from './vulnerabilities';
import * as organizations from './organizations';
import * as scans from './scans';
import * as users from './users';
import * as scanTasks from './scan-tasks';
import * as stats from './stats';
import { listenForDockerEvents } from './docker-events';
import { createProxyMiddleware } from 'http-proxy-middleware';

if (
  (process.env.IS_OFFLINE || process.env.IS_LOCAL) &&
  typeof jest === 'undefined'
) {
  listenForDockerEvents();

  setInterval(() => scheduler({}, {} as any, () => null), 30000);
}

const handlerToExpress = (handler) => async (req, res, next) => {
  const { statusCode, body } = await handler(
    {
      pathParameters: req.params,
      requestContext: req.requestContext,
      body: JSON.stringify(req.body || '{}'),
      headers: req.headers,
      path: req.originalUrl
    },
    {}
  );
  try {
    const parsedBody = JSON.parse(body);
    res.status(statusCode).json(parsedBody);
  } catch (e) {
    // Not a JSON body
    res.setHeader('content-type', 'text/plain');
    res.status(statusCode).send(body);
  }
};

const app = express();

app.use(cors());
app.use(bodyParser.json());
app.use(helmet.hsts());
app.use(cookieParser());

app.get('/', handlerToExpress(healthcheck));
app.post('/auth/login', handlerToExpress(auth.login));
app.post('/auth/callback', handlerToExpress(auth.callback));

const checkUserLoggedIn = async (req, res, next) => {
  req.requestContext = {
    authorizer: await auth.authorize({
      authorizationToken: req.headers.authorization
    })
  };
  if (
    !req.requestContext.authorizer.id ||
    req.requestContext.authorizer.id === 'cisa:crossfeed:anonymous'
  ) {
    return res.status(401).send('Not logged in');
  }
  return next();
};

const checkUserSignedTerms = (req, res, next) => {
  // Bypass ToU for CISA emails
  const approvedEmailAddresses = ['@cisa.dhs.gov'];
  if (process.env.NODE_ENV === 'test')
    approvedEmailAddresses.push('@crossfeed.cisa.gov');
  for (const email of approvedEmailAddresses) {
    if (
      req.requestContext.authorizer.email &&
      req.requestContext.authorizer.email.endsWith(email)
    )
      return next();
  }
  if (
    !req.requestContext.authorizer.dateAcceptedTerms ||
    (req.requestContext.authorizer.acceptedTermsVersion &&
      req.requestContext.authorizer.acceptedTermsVersion !==
        getToUVersion(req.requestContext.authorizer))
  ) {
    return res.status(403).send('User must accept terms of use');
  }
  return next();
};

const getMaximumRole = (user) => {
  if (user?.userType === 'globalView') return 'user';
  return user && user.roles && user.roles.find((role) => role.role === 'admin')
    ? 'admin'
    : 'user';
};

const getToUVersion = (user) => {
  return `v${process.env.REACT_APP_TERMS_VERSION}-${getMaximumRole(user)}`;
};

// Rewrite the URL for some Matomo admin dashboard URLs, due to a bug in
// how Matomo handles relative URLs when hosted on a subpath.
app.get('/plugins/Morpheus/images/logo.svg', (req, res) =>
  res.redirect('/matomo/plugins/Morpheus/images/logo.svg?matomo')
);
app.get('/index.php', (req, res) => res.redirect('/matomo/index.php'));

const matomoProxy = createProxyMiddleware({
  target: process.env.MATOMO_URL,
  headers: { HTTP_X_FORWARDED_URI: '/matomo' },
  pathRewrite: function (path, req) {
    return path.replace(/^\/matomo/, '');
  },
  onProxyReq: function (proxyReq, req, res) {
    // Only pass the MATOMO_SESSID cookie to Matomo.
    const cookies = cookie.parse(proxyReq.getHeader('Cookie'));
    const newCookies = cookie.serialize(
      'MATOMO_SESSID',
      String(cookies['MATOMO_SESSID'])
    );
    proxyReq.setHeader('Cookie', newCookies);
  },
  onProxyRes: function (proxyRes, req, res) {
    // Remove transfer-encoding: chunked responses, because API Gateway doesn't
    // support chunked encoding.
    if (proxyRes.headers['transfer-encoding'] === 'chunked') {
      proxyRes.headers['transfer-encoding'] = '';
    }
  }
});

app.use(
  '/matomo',
  async (req, res, next) => {
    // Public paths -- see https://matomo.org/docs/security-how-to/
    const ALLOWED_PATHS = ['/matomo.php', '/matomo.js'];
    if (ALLOWED_PATHS.indexOf(req.path) > -1) {
      return next();
    }
    // API Gateway isn't able to proxy fonts properly -- so we're using a CDN instead.
    if (req.path === '/plugins/Morpheus/fonts/matomo.woff2') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.woff2'
      );
    }
    if (req.path === '/plugins/Morpheus/fonts/matomo.woff') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.woff'
      );
    }
    if (req.path === '/plugins/Morpheus/fonts/matomo.ttf') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.ttf'
      );
    }
    // Only allow global admins to access all other paths.
    const user = (await auth.authorize({
      authorizationToken: req.cookies['crossfeed-token']
    })) as auth.UserToken;
    if (user.userType !== 'globalAdmin') {
      return res.status(401).send('Unauthorized');
    }
    return next();
  },
  matomoProxy
);

// Routes that require an authenticated user, without
// needing to sign the terms of service yet
const authenticatedNoTermsRoute = express.Router();
authenticatedNoTermsRoute.use(checkUserLoggedIn);
authenticatedNoTermsRoute.get('/users/me', handlerToExpress(users.me));
authenticatedNoTermsRoute.post(
  '/users/me/acceptTerms',
  handlerToExpress(users.acceptTerms)
);
authenticatedNoTermsRoute.put('/users/:userId', handlerToExpress(users.update));

app.use(authenticatedNoTermsRoute);

// Routes that require an authenticated user that has
// signed the terms of service
const authenticatedRoute = express.Router();
authenticatedRoute.use(checkUserLoggedIn);
authenticatedRoute.use(checkUserSignedTerms);

authenticatedRoute.post('/search', handlerToExpress(search.search));
authenticatedRoute.post('/domain/search', handlerToExpress(domains.list));
authenticatedRoute.post('/domain/export', handlerToExpress(domains.export_));
authenticatedRoute.get('/domain/:domainId', handlerToExpress(domains.get));
authenticatedRoute.post(
  '/vulnerabilities/search',
  handlerToExpress(vulnerabilities.list)
);
authenticatedRoute.post(
  '/vulnerabilities/export',
  handlerToExpress(vulnerabilities.export_)
);
authenticatedRoute.get(
  '/vulnerabilities/:vulnerabilityId',
  handlerToExpress(vulnerabilities.get)
);
authenticatedRoute.put(
  '/vulnerabilities/:vulnerabilityId',
  handlerToExpress(vulnerabilities.update)
);
authenticatedRoute.get('/scans', handlerToExpress(scans.list));
authenticatedRoute.get('/granularScans', handlerToExpress(scans.listGranular));
authenticatedRoute.post('/scans', handlerToExpress(scans.create));
authenticatedRoute.get('/scans/:scanId', handlerToExpress(scans.get));
authenticatedRoute.put('/scans/:scanId', handlerToExpress(scans.update));
authenticatedRoute.delete('/scans/:scanId', handlerToExpress(scans.del));
authenticatedRoute.post('/scans/:scanId/run', handlerToExpress(scans.runScan));
authenticatedRoute.post(
  '/scheduler/invoke',
  handlerToExpress(scans.invokeScheduler)
);
authenticatedRoute.post('/scan-tasks/search', handlerToExpress(scanTasks.list));
authenticatedRoute.post(
  '/scan-tasks/:scanTaskId/kill',
  handlerToExpress(scanTasks.kill)
);
authenticatedRoute.get(
  '/scan-tasks/:scanTaskId/logs',
  handlerToExpress(scanTasks.logs)
);

authenticatedRoute.get('/organizations', handlerToExpress(organizations.list));
authenticatedRoute.get(
  '/organizations/public',
  handlerToExpress(organizations.listPublicNames)
);
authenticatedRoute.get(
  '/organizations/:organizationId',
  handlerToExpress(organizations.get)
);
authenticatedRoute.post(
  '/organizations',
  handlerToExpress(organizations.create)
);
authenticatedRoute.put(
  '/organizations/:organizationId',
  handlerToExpress(organizations.update)
);
authenticatedRoute.delete(
  '/organizations/:organizationId',
  handlerToExpress(organizations.del)
);
authenticatedRoute.post(
  '/organizations/:organizationId/roles/:roleId/approve',
  handlerToExpress(organizations.approveRole)
);
authenticatedRoute.post(
  '/organizations/:organizationId/roles/:roleId/remove',
  handlerToExpress(organizations.removeRole)
);
authenticatedRoute.post(
  '/organizations/:organizationId/granularScans/:scanId/update',
  handlerToExpress(organizations.updateScan)
);
authenticatedRoute.post('/stats', handlerToExpress(stats.get));
authenticatedRoute.get('/users', handlerToExpress(users.list));
authenticatedRoute.post('/users', handlerToExpress(users.invite));
authenticatedRoute.delete('/users/:userId', handlerToExpress(users.del));

app.use(authenticatedRoute);

export default app;
