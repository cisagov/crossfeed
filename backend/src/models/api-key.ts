import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    // length based on hex formatted sha256
    length: 64,
    type: 'varchar'
  })
  hashedKey: string;

  @Column({
    length: 4,
    type: 'varchar'
  })
  lastFour: string;
}
