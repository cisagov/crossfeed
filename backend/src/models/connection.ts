import { createConnection, Connection } from 'typeorm';
import {
  Domain,
  Service,
  Vulnerability,
  Scan,
  Organization,
  User,
  Role,
  ScanTask
} from '.';

let connection: Connection | null = null;

const connectDb = async (logging?: boolean) => {
  const connection = createConnection({
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT ?? ''),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    entities: [
      Domain,
      Service,
      Vulnerability,
      Scan,
      Organization,
      User,
      Role,
      ScanTask
    ],
    synchronize: false,
    name: 'default',
    dropSchema: false,
    logging: logging ?? false
  });
  return connection;
};

export const connectToDatabase = async (logging?: boolean) => {
  if (!connection?.isConnected) {
    connection = await connectDb(logging);
  }
  if (typeof jest === 'undefined') {
    console.log('=> DB Connected');
  }
  return connection;
};
