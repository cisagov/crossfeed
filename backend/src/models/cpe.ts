import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  Unique
} from 'typeorm';
import { Cve } from './cve';

@Entity()
@Unique(['name', 'version', 'vendor'])
export class Cpe extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  name: string;

  @Column()
  version: string;

  @Column()
  vendor: string;

  @Column()
  lastSeenAt: Date;

  @ManyToMany(() => Cve, (cve) => cve.cpes)
  cves: Cve[];
}
