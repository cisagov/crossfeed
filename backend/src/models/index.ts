export * from './domain';
export * from './service';
export * from './connection';
export * from './vulnerability';
export * from './scan';
export * from './organization';
export * from './user';
export * from './role';
export * from './scan-task';
