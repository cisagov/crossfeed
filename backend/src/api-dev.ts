// Main entrypoint for dev server.

import app from './api/app';

process.env.IS_OFFLINE = 'true';

const port = 3000;
app.listen(port, () => {
  console.log('App listening on port ' + port);
});
