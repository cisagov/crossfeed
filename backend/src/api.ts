// Main entrypoint for serverless API code.

import * as serverless from 'serverless-http';
import app from './api/app';

module.exports.handler = serverless(app, {
  binary: ['image/*', 'font/*']
});
