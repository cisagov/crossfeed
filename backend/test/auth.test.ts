import * as request from 'supertest';
import app from '../src/api/app';
import { User } from '../src/models';
jest.mock('../src/api/login-gov');

jest.mock('jsonwebtoken', () => ({
  verify: (a, b, cb) => {
    cb(null, {
      email: a.split('TOKEN_')[1],
      email_verified: true,
      sub: Math.random() + ''
    });
  },
  sign: jest.requireActual('jsonwebtoken').sign
}));

describe('auth', () => {
  describe('login', () => {
    it('success', async () => {
      const response = await request(app).post('/auth/login').expect(200);
      expect(response.body).toMatchSnapshot();
    });
  });
  describe('callback', () => {
    it('success with login.gov', async () => {
      const response = await request(app)
        .post('/auth/callback')
        .send({
          code: 'CODE',
          state: 'STATE',
          origState: 'ORIGSTATE',
          nonce: 'NONCE'
        })
        .expect(200);
      expect(response.body.token).toBeTruthy();
      expect(response.body.user).toBeTruthy();
      expect(response.body.user.email).toEqual('test@crossfeed.cisa.gov');
      const user = await User.findOne(response.body.user.id);
      expect(user?.firstName).toEqual('');
      expect(user?.lastName).toEqual('');
    });
    it('success with cognito', async () => {
      process.env.USE_COGNITO = '1';
      const response = await request(app)
        .post('/auth/callback')
        .send({
          token: 'TOKEN_test2@crossfeed.cisa.gov'
        })
        .expect(200);
      expect(response.body.token).toBeTruthy();
      expect(response.body.user).toBeTruthy();
      expect(response.body.user.email).toEqual('test2@crossfeed.cisa.gov');
      const user = await User.findOne(response.body.user.id);
      expect(user?.firstName).toEqual('');
      expect(user?.lastName).toEqual('');
      process.env.USE_COGNITO = '';
    });

    it('success with cognito with two different ids - should overwrite cognitoId', async () => {
      process.env.USE_COGNITO = '1';
      let response = await request(app)
        .post('/auth/callback')
        .send({
          token: 'TOKEN_test3@crossfeed.cisa.gov'
        })
        .expect(200);
      expect(response.body.token).toBeTruthy();
      expect(response.body.user).toBeTruthy();
      expect(response.body.user.email).toEqual('test3@crossfeed.cisa.gov');
      const { id, cognitoId } = response.body.user;
      expect(cognitoId).toBeTruthy();
      expect(id).toBeTruthy();

      response = await request(app)
        .post('/auth/callback')
        .send({
          token: 'TOKEN_test3@crossfeed.cisa.gov'
        })
        .expect(200);
      const user = (await User.findOne(response.body.user.id)) as User;
      expect(user.id).toEqual(id);
      expect(user.cognitoId).not.toEqual(cognitoId);
      process.env.USE_COGNITO = '';
    });

    it('login with login.gov and later cognito should preserve ids', async () => {
      let response = await request(app)
        .post('/auth/callback')
        .send({
          code: 'CODE',
          state: 'STATE',
          origState: 'ORIGSTATE',
          nonce: 'NONCE'
        })
        .expect(200);
      expect(response.body.token).toBeTruthy();
      expect(response.body.user).toBeTruthy();
      expect(response.body.user.email).toEqual('test@crossfeed.cisa.gov');
      const { id, loginGovId } = response.body.user;
      expect(id).toBeTruthy();
      expect(loginGovId).toBeTruthy();

      process.env.USE_COGNITO = '1';
      response = await request(app)
        .post('/auth/callback')
        .send({
          token: 'TOKEN_test@crossfeed.cisa.gov'
        })
        .expect(200);
      expect(response.body.token).toBeTruthy();

      const user = (await User.findOne(response.body.user.id)) as User;
      expect(user.email).toEqual('test@crossfeed.cisa.gov');
      expect(user.id).toEqual(id);
      expect(user.loginGovId).toEqual(loginGovId);
      expect(user.cognitoId).toBeTruthy();
      process.env.USE_COGNITO = '';
    });
  });
});
