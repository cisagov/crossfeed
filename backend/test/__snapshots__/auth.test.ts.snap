// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth login success 1`] = `
Object {
  "nonce": "3ac04a92b121fce3fc5ca95251e70205d764147731",
  "redirectUrl": "https://idp.int.identitysandbox.gov/openid_connect/authorize?client_id=urn%3Agov%3Agsa%3Aopenidconnect.profiles%3Asp%3Asso%3Acisa%3Acrossfeed&scope=openid%20email&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%2F&acr_values=http%3A%2F%2Fidmanagement.gov%2Fns%2Fassurance%2Fial%2F1&nonce=3ac04a92b121fce3fc5ca95251e70205d764147731&state=258165a7dc37ce6034e74815e991606d49fd52240e469f&prompt=select_account",
  "state": "258165a7dc37ce6034e74815e991606d49fd52240e469f",
}
`;
